### DISCUSSION TITLE
Develop branch workflow and automations

### DISCUSSION THREAD

**Discussion by DhivinX** [xeokit/sdk#15] [Ideas]
Hello!

We could add Github Actions configuration to automate the publishing of packages when changes are pushed to the main branch. You would need to add a develop branch for the current code and a main branch would be the alive branch. Then when we merge the develop branch into main it would automatically publish the packages to the registry.

Then the workflow looks like this:

- People create their ```feature``` or ```fix``` branch from ```develop``` branch (ex. feat/register-form)
- After committing the changes, they create a PR with a return to ```develop```
- At the next milestone, we merge changes from ```develop``` to ```main```
- Then the Github Actions workflow starts, which lints, tests, builds the package and publishes it to the registry

We could set the ```main``` branch as a protected branch, so that no one can commit directly to ```main```, only PR from ```develop```.

What do you think?
Please share any thoughts.

**Comment by xeolabs** [comment]
Hi @DhivinX, yes that works perfectly.

**Comment by xeolabs** [comment]
I've made a `develop` branch and protected the `main` branch.

**Comment by xeolabs** [comment]
When generating typedocs, should they be generated from each package's `./src` or from `./dist`?

**Reply by DhivinX** [reply]
IMO from ```./src```

**Reply by xeolabs** [reply]
Yip I agree.

**Comment by DhivinX** [comment]
@xeolabs I saw you started some fix for tsconfigs. Generally, I wanted to create two packages in the repo, if possible, in ```packages```, ```tsconfig``` and ```config```, one containing the general tsconfig from which all others will be extended, and the other containing the eslint/prettier configurations. Is this idea right for you?

**Comment by xeolabs** [comment]
> IMO from `./src`



> @xeolabs I saw you started some fix for tsconfigs. Generally, I wanted to create two packages in the repo, if possible, in `packages`, `tsconfig` and `config`, one containing the general tsconfig from which all others will be extended, and the other containing the eslint/prettier configurations. Is this idea right for you?

Yes this sounds good to me.

**Comment by xeolabs** [comment]
Two more points occurred to me:

- We'll need some way to cause some packages to not get deployed, since there are some packages that are ready to deploy, and some that are incomplete.I guess we can just remove the `publish` task from their `package.json` configs until they're ready, unless there is a cleaner way
- Perhaps we could use changesets, to manage easier publication of the multiple packages
- Jest test reports - if possible, it would be cool if we could have a single uber HTML test report page for all packages in the monorepo, that way the health of the whole set could be checked at a glance

**Reply by DhivinX** [reply]
1. Yes, we can exclude packages for publishing using .npmignore, I added such a file in the previous PR (the syntax is the same as .gitignore), I just couldn't test if it works well
2. Chnagesets is a very good idea, but I see it is already implemented
3. Ok, I understand, first I will fix the problems with test execution and then I will try to force the collective test to be generated in HTML, if possible

**Comment by paireks** [comment]
Hey guys, I have one question, cause I get the idea of having develop branch and protecting the main one, but could you please explain to me the reason of switching default branch into develop? Shouldn't the main branch still be the default one?

**Comment by xeolabs** [comment]
As I understand it, "develop" is the default branch because when contributors want to branch-and-PR they would be making the branch off "develop".

Also, "main" wont contain anything until there is releasable code in "develop" to merge into it.

**Reply by paireks** [reply]
Ok, I see, thank you for explanaition, it makes sense. I was more familiar with still keeping main branches as default, having contributors branch-and-PR from main, but into the develop, that's why I was confused at first :)

**Comment by xeolabs** [comment]
Hi @DhivinX I'm having trouble getting things to build and I can't figure out what's not working. Could you please take a look?


I'm doing a fresh clone, then checking out the `develop` branch, and running the build instructions in the README, but when I do `pnpm install` I get these messages:

````
lindsay@mousebear:~/xeolabs/sdk-jul28$ pnpm install
Scope: all 44 workspace projects
/home/lindsay/xeolabs/sdk-jul28/packages/bcf:
 ERR_PNPM_FETCH_404  GET https://registry.npmjs.org/@xeokit%2Ftestutils: Not Found - 404

This error happened while installing a direct dependency of /home/lindsay/xeolabs/sdk-jul28/packages/bcf

@xeokit/testutils is not in the npm registry, or you have no permission to fetch it.

An authorization header was used: Bearer npm_[hidden]
Progress: resolved 37, reused 36, downloaded 0, added 0
lindsay@mousebear:~/xeolabs/sdk-jul28$ 
````

If you can see anything that stands out, that would be much appreciated. AKAIK I can't see anything I broke with the configs etc..

**Reply by DhivinX** [reply]
Hi @xeolabs I'll check this.

**Reply by DhivinX** [reply]
@xeolabs please check #44

**Comment by xeolabs** [comment]
Nice, that working now, thanks @DhivinX

**Reply by DhivinX** [reply]
You're welcome. If you need anything, let me know. 😉
